Y25-296 Import Kinnex samples

seenanair opened the pull request:
Closes https://github.com/sanger/traction-ui/issues/2316

#### Changes proposed in this pull request

This pull request introduces support for handling "Sequencescape Kinnex Tubes" in the reception pipeline.

Kinnex tubes contain multiple samples that share the same supplier_name. A compound sample is created using this supplier_name, which is stored as the sample_name and used as the key.

Key changes include adding a new module for processing Kinnex Tubes, updating configurations to integrate this new type into the system, and implementing corresponding tests to ensure functionality.

### New Feature: Support for Sequencescape Kinnex Tubes
* Added `src/lib/receptions/SequencescapeKinnexTubes.js` to handle fetching and transforming labware data for Kinnex Tubes, including `fetchLabwareForReception` and `getAttributeKeys` methods. 
* Updated `src/lib/receptions/index.js` to include Sequencescape Kinnex Tubes in `ReceptionTypes` and `Receptions`. 
* Added `buildKinnexRequestAndSample` and `transformKinnexTube` functions to `src/lib/receptions/sequencescapeUtils.js` for request and sample object creation and data transformation. 
* Introduced a new `kinnex_tubes` type in `labwareTypes` in `src/lib/receptions/sequencescapeUtils.js`

Below is a UI screenshot demonstrating the newly added Kinnex Tubes import feature in the Reception page.

![Screenshot 2025-06-05 at 13 57 09](https://github.com/user-attachments/assets/e555e1dd-7ebe-4901-85db-d9ce53f2f993)

#### Instructions for Reviewers
1. Kinnex Methods
It was agreed during the discussion that it’s beneficial to keep the logic for fetching and transforming labware data for Kinnex Tubes as separate methods and utility functions.
Though this introduces some code repetition, the added clarity and modularity are considered worthwhile trade-offs.

2. Attributes 
   - Attribute Mapping : The following attributes are assumed to be equivalent:
      a. Sample name → supplier_sample_name (All samples in a Kinnex tube share the same supplier_name, which serves    as a unique identifier for the compound sample created in Traction to represent all samples in the tube.)
      b. Sample type → donor_id (represents donor information per sample)
      c. Study number / Study name → study_id
      d. Date submitted → sample_collection_date
  
   - Attributes Requiring Further Clarification: The following attributes still need input or confirmation from stakeholders:
      
      Sample ID
      Cohort
      Cell type and Number of requested cells

3. Sequencescape PR
     A SS [PR](https://github.com/sanger/sequencescape/pull/4985) is created which introduces a new attribute, date_of_sample_collection, to the SampleMetadataResource class

stevieing commented on tests/e2e/specs/reception/import_samples_from_sequencescapeKinnexTubes.cy.js:
```diff
@@ -0,0 +1,152 @@
+      })
+    })
+  })
+  describe('Successfully - V2', () => {
```
What does rhe V2 mean?

seenanair replied:
I think it perhaps refers to SS V2. I will change it to 'Importing samples from Sequencescape Kinnex Tubes'

seenanair commented on src/lib/receptions/SequencescapeKinnexTubes.js:
```diff
@@ -0,0 +1,56 @@
+import { fetchLabwareFromSequencescape, labwareTypes } from './sequencescapeUtils.js'
+// This module replaces by services.Sequencescape which can be removed
```
This is copied and pasted from another similar file. Sorry, I forgot to remove the comment – I’m not sure what it means.

BenTopping commented on tests/e2e/specs/reception/import_samples_from_sequencescapeKinnexTubes.cy.js:
```diff
@@ -0,0 +1,155 @@
+      cy.intercept('POST', '/api/scans', {
+        statusCode: 201,
+        body: {
+          message: 'SE108532I successfully stored in LRT006 Draw 1',
```
Should this be "DN9000002A" or "NT10" instead of "SE108532I"?

BenTopping commented on tests/unit/views/GeneralReception.spec.js:
```diff
@@ -100,6 +101,11 @@ describe('GeneralReception', () => {
       expect(wrapper.find('[data-type=pipeline-list]').findAll('option')[0].text()).toBe('ONT')
     })
+    it('only shows Pacbio for Sequencescape Kinnex Tubes', async () => {
+      await wrapper.find('[data-type=source-list]').findAll('option')[3].setSelected()
```
Probably better to find the option from the name rather than the index as the indexes could change, but I can see its done that way for the other tests.

seenanair replied:
Good suggestion. Done 👍

seenanair commented:
@stevieing @BenTopping Sorry, the PR is again updated.
The changes are:
- Updated reception for compound samples so the UI sends an array of samples to the service instead of a single compound sample
- Added a feature flag to enable/disable Kinnex Tube reception
- Mocked Kinnex compound sample reception for testing purposes

BenTopping commented on src/lib/receptions/SequencescapeKinnexTubes.js:
```diff
@@ -40,7 +40,7 @@ const fetchLabwareForReception = async ({ requests, barcodes, requestOptions })
  * @returns {Array} Array of attribute keys
  */
-const getAttributeKeys = () => ['tubes_attributes']
+const getAttributeKeys = () => ['compound_sample_tubes_attributes']
```
Where is this function used, can't see where its used from a brief look at the code base, same goes for the other receptions with this function.

seenanair replied:
This function is set up in the Receptions config and is accessed dynamically depending on which reception type is being used which helps component to figure out which attribute keys to use for each workflow, allowing correct attributes to be retrieved for each reception type.

BenTopping commented on src/lib/receptions/MockReception.js:
```diff
@@ -25,6 +25,30 @@ const buildMockSampleAndRequest = (sample_name, requestOptions) => {
+ * @param {Object} requestOptions - An object containing request options to be included in each well's request.
+ * @returns {Object} A mock sample and request object.
+ */
+const buildMockCompoundSampleAndRequest = (sample_names, requestOptions) => {
```
Since this mostly a copy of `buildMockSampleAndRequest` but as a list, could they be combined or could this be removed?

seenanair replied:
Good point, It is good to refactor `buildMockCompoundSampleAndRequest`  to use `buildMockSampleAndRequest` to create that list, so there wont be duplication.

BenTopping reviewed (commented):
Just two minor comments otherwise looks good.
I assume there has been a discussion to move the 'compounding' of the sample to the service and thats why we are now sending all samples?
